src/Sylius/Behat/Client/RequestFactory.php

```diff
@@ -0,0 +1,179 @@
+            HttpRequest::METHOD_PUT,
+        );
+        $builder->withHeader('CONTENT_TYPE', 'application/ld+json');
+        $builder->withHeader('HTTP_ACCEPT', 'application/ld+json');
+
+        if ($token) {
+            $builder->withHeader('HTTP_' . $authorizationHeader, 'Bearer ' . $token);
+        }
+
+        return $builder->build();
+    }
+
+    public function delete(
+        string $section,
+        string $resource,
+        string $id,
+        string $authorizationHeader,
+        ?string $token = null
+    ): RequestInterface {
+        $builder = RequestBuilder::create(
+            sprintf('/api/v2/%s/%s/%s', $section, $resource, $id),
+            HttpRequest::METHOD_DELETE,
+        );
+        $builder->withHeader('HTTP_ACCEPT', 'application/ld+json');
+
+        if ($token) {
+            $builder->withHeader('HTTP_' . $authorizationHeader, 'Bearer ' . $token);
+        }
+
+        return $builder->build();
+    }
+
+    public function transition(string $section, string $resource, string $id, string $transition): RequestInterface
+    {
+        return self::customItemAction($section, $resource, $id, HttpRequest::METHOD_PATCH, $transition);
+    }
+
+    public function customItemAction(string $section, string $resource, string $id, string $type, string $action): RequestInterface
+    {
+        $builder = RequestBuilder::create(
+            sprintf('/api/v2/%s/%s/%s/%s', $section, $resource, $id, $action),
+            $type,
+        );
+        $builder->withHeader('CONTENT_TYPE', self::resolveHttpMethod($type));
+
+        return $builder->build();
+    }
+
+    public function upload(
+        string $section,
+        string $resource,
+        string $authorizationHeader,
+        ?string $token = null
+    ): RequestInterface {
+        $builder = RequestBuilder::create(
+            sprintf('/api/v2/%s/%s', $section, $resource),
+            HttpRequest::METHOD_POST,
+        );
+        $builder->withHeader('CONTENT_TYPE', 'multipart/form-data');
+        $builder->withHeader('HTTP_ACCEPT', 'application/ld+json');
+
+        if ($token) {
+            $builder->withHeader('HTTP_' . $authorizationHeader, 'Bearer ' . $token);
+        }
+
+        return $builder->build();
+    }
+//    Something like "custom" should not exist in my opinion
+    public static function custom(string $url, string $method, array $additionalHeaders = [], ?string $token = null): RequestInterface
+    {
+        $builder = RequestBuilder::create($url, $method);
+        $builder->withHeader('HTTP_ACCEPT', 'application/ld+json');
+        $builder->withHeader('CONTENT_TYPE', self::resolveHttpMethod($method));
+
+        if ($token) {
+            $builder->withHeader('HTTP_Authorization', 'Bearer ' . $token);
+        }
+
+        foreach ($additionalHeaders as $name => $value) {
+            $builder->withHeader($name, $value);
+        }
+
+        return $builder->build();
```

Comment:
I'm thinking if we shouldn't return the builder and allow for a build call in the contexts. This would allow for the decoration of this object as well. Perhaps two services would be a good choice? Then we could finalize this service and the default approach to extending it would be with decoration

Reply:
I like this idea, but I would rather stick with the current solution for now, and in another PR return Builder. To achieve that I will need to refactor a lot ApiPlatform Client

Reply:
We could introduce RequestBuilderFactory which will have almost the same API but will return builder, which will be called in factories. But perhaps I'm making it too complex.